File: C#/Relationship.cs
// Class relationships in C#.
//
//   Association - "uses-a" (independent lifetimes)
//   Aggregation - "has-a"  (whole holds references to parts that outlive it)
//   Composition - "owns-a" (whole creates and owns its parts)

using System;
using System.Collections.Generic;
using System.Linq;

namespace LLDWeekend4.Relationships
{
    // ------------------------------------------------------------
    // Association
    // ------------------------------------------------------------
    public class Student
    {
        public string Name { get; }
        public Student(string name) => Name = name;
    }

    public class Payment
    {
        public double Amount { get; }
        public Payment(double amount) => Amount = amount;
    }

    public class Teacher
    {
        public void Teaches(IEnumerable<Student> students)
        {
            foreach (var student in students)
                Console.WriteLine($"Teaching {student.Name}");
        }

        public void GetSalary(Payment payment)
            => Console.WriteLine($"Teacher received {payment.Amount}");
    }

    // ------------------------------------------------------------
    // Aggregation (students exist outside the course)
    // ------------------------------------------------------------
    public class LldCourse
    {
        private readonly List<Student> _students = new();   // references, not owned

        public void Enroll(Student student) => _students.Add(student);

        public int Count => _students.Count;
    }

    // ------------------------------------------------------------
    // Composition (a directory owns its children)
    // ------------------------------------------------------------
    public interface IFileSystemNode
    {
        string Name { get; }
    }

    public class File : IFileSystemNode
    {
        public string Name { get; }
        private readonly string _meta;

        public File(string name, string meta)
        {
            Name = name;
            _meta = meta;
        }
    }

    public class Directory : IFileSystemNode
    {
        public string Name { get; }
        private readonly List<IFileSystemNode> _children = new();  // owned

        public Directory(string name) => Name = name;

        public void AddChild(string name, string meta)
            => _children.Add(new File(name, meta));

        public IReadOnlyList<IFileSystemNode> Children => _children;
    }

    // ------------------------------------------------------------
    // Runner
    // ------------------------------------------------------------
    public static class RelationshipDemo
    {
        public static void Run()
        {
            var students = new List<Student> { new("Neha"), new("Rahul") };

            var teacher = new Teacher();
            teacher.Teaches(students);
            teacher.GetSalary(new Payment(50000));

            var course = new LldCourse();
            foreach (var student in students) course.Enroll(student);
            Console.WriteLine($"Course has {course.Count} students");

            var root = new Directory("root");
            root.AddChild("Main.cs", "text/x-csharp");
            Console.WriteLine(
                $"{root.Name} contains {string.Join(", ", root.Children.Select(c => c.Name))}");
        }
    }
}

File: TypeScript/Relationship.ts
/*
 * Class relationships in TypeScript.
 *
 *   Association - "uses-a" (independent lifetimes)
 *   Aggregation - "has-a"  (whole references parts that outlive it)
 *   Composition - "owns-a" (whole creates and owns its parts)
 */

// ------------------------------------------------------------
// Association
// ------------------------------------------------------------
class Student {
  constructor(public readonly name: string) {}
}

class Payment {
  constructor(public readonly amount: number) {}
}

class Teacher {
  teaches(students: Student[]): void {
    for (const student of students) {
      console.log(`Teaching ${student.name}`);
    }
  }

  getSalary(payment: Payment): void {
    console.log(`Teacher received ${payment.amount}`);
  }
}

// ------------------------------------------------------------
// Aggregation (students exist outside the course)
// ------------------------------------------------------------
class LLDCourse {
  private readonly students: Student[] = []; // references, not owned

  enroll(student: Student): void {
    this.students.push(student);
  }

  get size(): number {
    return this.students.length;
  }
}

// ------------------------------------------------------------
// Composition (a directory owns its children)
// ------------------------------------------------------------
interface FileSystemNode {
  readonly name: string;
}

class FileNode implements FileSystemNode {
  constructor(
    public readonly name: string,
    private readonly meta: string,
  ) {}
}

class Directory implements FileSystemNode {
  private readonly childNodes: FileSystemNode[] = []; // owned

  constructor(public readonly name: string) {}

  addChild(name: string, meta: string): void {
    this.childNodes.push(new FileNode(name, meta));
  }

  get children(): readonly FileSystemNode[] {
    return this.childNodes;
  }
}

function main(): void {
  const students = [new Student('Neha'), new Student('Rahul')];

  const teacher = new Teacher();
  teacher.teaches(students);
  teacher.getSalary(new Payment(50000));

  const course = new LLDCourse();
  students.forEach((s) => course.enroll(s));
  console.log(`Course has ${course.size} students`);

  const root = new Directory('root');
  root.addChild('Main.ts', 'text/typescript');
  console.log(`${root.name} contains ${root.children.map((c) => c.name).join(', ')}`);
}

main();

export {};
